Gemini AI translation option

destpstrzy opened the pull request:
**- Gemini translating option added**

1. Translation module is a modified version of https://github.com/MaKTaiL/gemini-srt-translator
2. You can find this option on subitle options in movie/episode view
3. You need to set your API key first in settings in "Subtitles" section first
4. You can check option for adding "# Subtitles translated with Gemini # " at the beginning of the subtitles

destpstrzy commented:
I see that checks were not successful. Could you help me what should I fix?

morpheus65535 commented:
18 new vendored modules? Only 11 in versions.txt? That's quite a lot to add. I'm not event sure I agree to add more, we're trying to reduce the number of dependencies.

Python 3.8 must be used to install dependencies. Cleanup must be done after that.
```
python3.8 -m pip install --upgrade MOMDULE_NAME --no-deps --no-compile --no-binary=:all: --no-cache-dir -t /Users/morpheus/Desktop/bazarr/libs/
rm -R -f /Users/morpheus/Desktop/bazarr/libs/bin
rm -R -f /Users/morpheus/Desktop/bazarr/libs/docs
rm -f /Users/morpheus/Desktop/bazarr/libs/CHANGELOG.md
rm -f /Users/morpheus/Desktop/bazarr/libs/README.md
find /Users/morpheus/Desktop/bazarr/libs/* -name '*.so' -delete
```
Is it possible to have all translation providers under a sub menu?

morpheus65535 commented:
It would have been a good idea to come discuss this before putting that much time in a PR.

morpheus65535 commented:
> I see that checks were not successful. Could you help me what should I fix?

```
error during build:
[commonjs--resolver] node_modules/npm/lib/npm.js (372:26): Legacy octal escape is not permitted in strict mode
file: /home/runner/work/bazarr/bazarr/frontend/node_modules/npm/lib/npm.js:372:26 (/home/runner/work/bazarr/bazarr/frontend/node_modules/latest/index.js)
```

destpstrzy commented:
> It would have been a good idea to come discuss this before putting that much time in a PR.

Sorry about that. I have just tested gemini translated and was impressed with the results and had some free time to play with integration. I should ask first.

I can integrate it with 'Translate...' button and will try to reduce dependencies as much as I can if that would be ok

destpstrzy commented:
Ok, I have tried do some changes but the dependencies are from Gemini API and I dont have a skill to workaround this :( So if it is too much dependencies I can just abandon this for now.

anderson-oki commented:
Consuming the Gemini rest api would remove all dependencies and its not that hard, but to be honest I think i would had integrated the options that we have on deep translator package since we are already using, i am sure gpt would catch up to gemini

morpheus65535 commented:
I would happily accept a PR that implement other providers from deep-translator! Like @anderson-oki wrote, I'm pretty sure there's other options in there that would give really nice result as well!

destpstrzy commented:
Thank you for the suggestions. I’m trying to use the REST API. I think this Gemini library produces good results because it makes effective use of system_instruction, sets the AI’s role properly, and allows sending large chunks of subtitles to preserve context.
I’m going to start with that and will try to implement the rest using Deep Translator.
I don’t use the languages from this project on a daily basis (I’m more of a Bazarr enthusiast and wanted to add some features), so it might take me some time.

anderson-comvex commented on frontend/package.json:
```diff
@@ -23,6 +23,7 @@
     "@tanstack/react-table": "^8.19.2",
     "axios": "^1.8.2",
     "braces": "^3.0.3",
+    "latest": "^0.2.0",
```
What is this package here?

anderson-comvex replied:
What is this package here?

anderson-comvex commented:
There are still a bunch of libs changes in the branch

anderson-comvex commented:
I also see a bunch of weird changes like get latest version from pypi?

Custom progress bar? What is going on here?

destpstrzy commented:
> I also see a bunch of weird changes like get latest version from pypi?
> 
> Custom progress bar? What is going on here?

It was leftover from old library for debugging. I have deleted it now.
Only depndency I needed to add was json_repair.

destpstrzy commented:
@anderson-comvex Could you tell me what should I do for your request: "Requested changes with read-only permissions"?
Also is there any way to use CollapseBox for other setting value? I want to hide Gemini settings if it is not selected (and other translators in future).

Gylesie commented:
With all the unnecessary baggage around this PR, this feels like it was vibe-coded.

anderson-oki commented:
Where is Json repair even being used?

anderson-oki commented on frontend/package-lock.json:
Package lock still contains the weird change

destpstrzy replied:
Fixed. I was branching from dev and it was changed there I think.

morpheus65535 commented on libs/bin/filetype:
Should be removed

destpstrzy replied:
Done

morpheus65535 commented on frontend/src/pages/Settings/General/index.tsx:
Why did you moved that to Settings-->Subtitles?

destpstrzy replied:
Fixed

morpheus65535 commented on libs/version.txt:
Cleanup must be done here

destpstrzy replied:
fixed

morpheus65535 commented:
> Only dependency I needed to add was json_repair.

And why not use normal JSON?

anderson-oki commented on frontend/src/types/settings.d.ts:
```diff
@@ -173,6 +173,13 @@ declare namespace Settings {
     excluded_tags: string[];
   }
 
+  interface Translating {
```
Let's rename this to `Translation` as a noun and not a present continuous verb.

destpstrzy replied:
Done

anderson-oki commented on frontend/src/types/settings.d.ts:
```diff
@@ -173,6 +173,13 @@ declare namespace Settings {
   }
 
+  interface Translating {
+    translator: string;
```
maybe translator_type?

```suggestion
    translator_type: string;
```

destpstrzy replied:
Done, thank you

destpstrzy commented on custom_libs/subliminal_patch/providers/napiprojekt.py:
```diff
@@ -135,7 +135,7 @@ def _scrape(self, video, language):
 
                     if self.only_authors:
-                        if author.lower() in ["brak", "automat", "si", "chatgpt", "ai", "robot", "maszynowe", "tłumaczenie maszynowe"]:
+                        if author.lower() in ["brak", "automat", "si", "chatgpt", "ai", "deepl", "robot", "maszynowe", "translator" "tłumaczenie maszynowe"]:
```
It is "Small improvements for Napiprojekt provider AI detection"

anderson-comvex replied:
I think the question was more like why this was included on this PR

anderson-oki replied:
Sorry i meant why this ended up being included on this PR.

destpstrzy commented on frontend/src/types/settings.d.ts:
```diff
@@ -173,6 +173,13 @@ declare namespace Settings {
     excluded_tags: string[];
   }
 
+  interface Translating {
```
Done

destpstrzy commented:
Thank you for your help. I make more cleaning and removed json_repair dependency. I have also added gettin movie/series description for AI translating context. It still shows version.txt changed but I don't know why.

destpstrzy commented:
I have an update for better handling of the new Gemini 2.5 Flash, which should now translate everything in a single request. I left the 2.0 flash model as default because it is currently the stable version

Do you think that it is good now for merging?

In the future after testing I could implement other LLMs. Sending audio to gemini could be interesting too.

destpstrzy reviewed (commented):
Done

morpheus65535 commented:
Under Python 3.8 (current minimum Python version):

```
Traceback (most recent call last):
  File "/Users/morpheus/Desktop/bazarr/bazarr/main.py", line 43, in <module>
    from app.server import webserver, app  # noqa E402
  File "/Users/morpheus/Desktop/bazarr/bazarr/app/server.py", line 13, in <module>
    from api import api_bp
  File "/Users/morpheus/Desktop/bazarr/bazarr/api/__init__.py", line 13, in <module>
    from .subtitles import api_ns_list_subtitles
  File "/Users/morpheus/Desktop/bazarr/bazarr/api/subtitles/__init__.py", line 3, in <module>
    from .subtitles import api_ns_subtitles
  File "/Users/morpheus/Desktop/bazarr/bazarr/api/subtitles/subtitles.py", line 14, in <module>
    from subtitles.tools.translate import translate_subtitles_file
  File "/Users/morpheus/Desktop/bazarr/bazarr/subtitles/tools/translate.py", line 22, in <module>
    from subtitles.tools.translator_gemini import TranslatorGemini
  File "/Users/morpheus/Desktop/bazarr/bazarr/subtitles/tools/translator_gemini.py", line 30, in <module>
    class TranslatorGemini:
  File "/Users/morpheus/Desktop/bazarr/bazarr/subtitles/tools/translator_gemini.py", line 322, in TranslatorGemini
    batch: list[SubtitleObject],
TypeError: 'type' object is not subscriptable
```

morpheus65535 commented on bazarr/subtitles/tools/translator_gemini.py:
```diff
@@ -0,0 +1,449 @@
+
+        except Exception as e:
+            exit(0)
+            throw(e)
```
unreachable code?

destpstrzy replied:
I have removed this.

morpheus65535 commented on bazarr/subtitles/tools/translator_gemini.py:
Why all those exit()? You really want to exit Bazarr when you get an exception?

morpheus65535 commented:
> I have left warning because I dont know how to hide it if not Google Translate is set. Could you help me with that?

@anderson-oki could you guide us on how to hide the Google Translate warning in translate modal if Gemini is the selected translator provider?

destpstrzy commented:
> > Yes, it is the same and use translator set in the settings.
> 
> And what about the available languages in the translate modal menu? They are static? Gemini support exactly the same languages as Google Translate?

Language filtering could be disabled for trasnaltors other than "google". I have tried to implement this but I dont know how to get "trasnlator_type" setting to TrasnaltionForm.

anderson-oki commented:
> > I have left warning because I dont know how to hide it if not Google Translate is set. Could you help me with that?
> 
> @anderson-oki could you guide us on how to hide the Google Translate warning in translate modal if Gemini is the selected translator provider?

@morpheus65535 

The whole `TranslationForm.tsx` seems to be based and made for GoogleTranslate. There are a huge hardcoded list of languages.

You can disable the alert by doing `const settings = useSettings();` and check if the settings for translation is gemini, but i am not sure if that is the appropriated solution at this point. Would need more time to check.

anderson-oki commented:
Is the response coming from the api in the network tab?

destpstrzy commented:
Yes. I have tried it by adding:
```
  // Get translation service name from settings
  // eslint-disable-next-line no-console
  console.log("Complete Settings object:", settings);
```

anderson-oki commented:
Maybe add a screenshot of the response in the network tab. I asked 2 questions where you cant answer yes for both at the same time so i am not sure what Yes is referring to.

destpstrzy commented:
Ok, but response of what request?
In the console tab I get: Complete Settings object: null

anderson-oki commented:
Try using

```
  const settings = useSystemSettings();
```

destpstrzy commented:
Thank you. That method worked.
Translating form now:
- shows chosen translator and model
- filter languages only for Google Translate

![image](https://github.com/user-attachments/assets/92c8d1b8-b452-40d2-acf5-2b73ea00de8e)
![image](https://github.com/user-attachments/assets/64aa0515-3e38-4a70-816f-e89ef29c3fea)

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,41 @@ const TranslationForm: FunctionComponent<Props> = ({
-    () => languages.filter((v) => v.code2 in translations),
-    [languages],
-  );
+  const TranslatorType = settings?.data?.translator?.translator_type;
```
Camel case for variables, use translationType

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,41 @@ const TranslationForm: FunctionComponent<Props> = ({
-    [languages],
-  );
+  const TranslatorType = settings?.data?.translator?.translator_type;
+  const isGoogleTranslator = TranslatorType === "translate";
```
Why the type is translate? Shouldnt be google translate or gemini or something else? This is strange

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,41 @@ const TranslationForm: FunctionComponent<Props> = ({
   );
 
+  // Get translation service name from settings
+  // eslint-disable-next-line no-console
```
Do not use console.log and do not disable any linters.

 Check other logs and follow the pattern

anderson-oki replied:
Do not use console.log and do not disable any linters.

 Check other logs and follow the pattern

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,41 @@ const TranslationForm: FunctionComponent<Props> = ({
+  // eslint-disable-next-line no-console
+  console.log("Complete Settings object:", settings?.data);
+
+  const TranslatorModel =
```
Camel case

destpstrzy replied:
Fixed

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,41 @@ const TranslationForm: FunctionComponent<Props> = ({
+      : TranslatorType === "gemini"
+        ? " (" + settings?.data?.translator?.gemini_model + ")"
+        : "";
+  const TranslatorService =
```
Camel case

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,41 @@ const TranslationForm: FunctionComponent<Props> = ({
+        ? " (" + settings?.data?.translator?.gemini_model + ")"
+        : "";
+  const TranslatorService =
+    TranslatorType === "translate"
```
This logic is weird and redundant, why do you need to double check? Can you refactor please

anderson-oki commented on frontend/src/pages/Settings/Subtitles/options.ts:
```diff
@@ -182,3 +182,14 @@ export const syncMaxOffsetSecondsOptions: SelectorOption<number>[] = [
+export const translatorOption: SelectorOption<string>[] = [
+  {
+    label: "Google Translate",
+    value: "translate",
```
Can we use google_translate instead to be less confusing?

anderson-oki commented on frontend/src/components/forms/TranslationForm.tsx:
```diff
@@ -144,17 +146,44 @@ const TranslationForm: FunctionComponent<Props> = ({
     (v) => v.code2,
   );
 
+  const translatorConfig = {
+    googleTranslate: {
+      service: "Google Translate",
+      model: "",
+    },
+    gemini: {
+      service: "Gemini",
+      model: ` (${settings?.data?.translator?.gemini_model || ""})`,
+    },
+  } as const;
```
Its a small thing but i would rather refactor this to a method instead of doing all these casts

interface TranslationConfig {
  code: string,
  service: string
  model?: string | null
}

const getTranslationConfig = (settings) : TranslationConfig => {
  if (settings.data.translator.type === 'gemini') {
    return {
      service: "Gemini",
      model: ` (${settings?.data?.translator?.gemini_model || ""})`,
     }
  }

  return {
   code: 'google_translate',
    service: "Google Translate",
      model: "",
  }
}

destpstrzy replied:
Thank you. I have changed that.

morpheus65535 commented:
Thanks!

destpstrzy commented:
> I have been testing this heavily today, including the new hotfixes, and I have a suggestion: Try to save the progress if possible - sometimes after 10-15 minutes of translating a movie, gemini throws an error and the whole progress is lost. If we could keep the progress file and continue from there, that would save so many repeated requests.

Thank you for the feedback. What error do you get the most often? It could be changed to keep translation processing file but you would need to remember to remove it manually so I made it to delete it at the end.

destpstrzy commented:
@morpheus65535 
I have a question about tagging auto translated subtitle. Is there any way to do that.
For now when I do machine translation and the official subtitles will be out I don't get it downloaded and stuck with AI ones.
What is a workaround for me I change the translated language in filename (for example .ar.srt) and Kometa add overlay on poster "AI translated" when there is 'ar' subtitles and no my language subtitles. Co you think adding feature to change ai subtitles languages would be good option or it is too weird? Or maybe some option to still monitor translated subtitles?
